base/server/src/main/java/com/netscape/cms/realm/PKILDAPRealm.java

```diff
@@ -0,0 +1,415 @@
+                    if ("ldap".equals(protocol)) {
+                        secureConn = "false";
+                    } else if ("ldaps".equals(protocol)) {
+                        secureConn = "true";
+                    } else {
+                        throw new LifecycleException("Unsupported LDAP protocol: " + protocol);
+                    }
+
+                    ldapConfig.put("ldapconn.host", host);
+                    ldapConfig.put("ldapconn.port", port);
+                    ldapConfig.put("ldapconn.secureConn", secureConn);
+                    break;
+                case "authType":
+                    logger.info("- authentication type: " + value);
+                    ldapConfig.put("ldapauth.authtype", value);
+                    break;
+                case "bindDN":
+                    logger.info("- bind DN: " + value);
+                    ldapConfig.put("ldapauth.bindDN", value);
+                    break;
+                case "bindPassword":
+                    ldapConfig.put("ldapauth.bindPassword", value);
+                    break;
+                case "nickname":
+                    logger.info("- nickname: " + value);
+                    ldapConfig.put("ldapauth.clientCertNickname", value);
+                    break;
+                case "minConns","maxConns","maxResults", "errorIfDown":
+                    logger.info("- " + name + ": " + value);
+                    ldapConfig.put(name, value);
+                    break;
+                default:
+                    if(name.startsWith("ldapauth.") || name.startsWith("ldapconn.")) {
+                        logger.info("- " + name + ": " + value);
+                        ldapConfig.put(name, value);
+                    }
+                    else {
+                        cs.put(name, value);
+                    }
+                }
+            }
+
+        } else {
+
+            logger.info("Loading LDAP realm config from " + configFile);
+            try {
+                cs = new EngineConfig(new FileConfigStorage(configFile));
+                cs.load();
+
+                ps = IPasswordStore.create(cs.getPasswordStoreConfig());
+            } catch (Exception e) {
+                throw new LifecycleException("Cannot load config file");
+            }
+            ldapConfig = cs.getInternalDBConfig();
+        }
+
+        socketConfig = cs.getSocketConfig();
+        connConfig = ldapConfig.getConnectionConfig();
+        authConfig = ldapConfig.getAuthenticationConfig();
+
+        usersDN = config.getParameter("usersDN");
+        logger.info("- users DN: " + usersDN);
+
+        groupsDN = config.getParameter("groupsDN");
+        logger.info("- groups DN: " + groupsDN);
+
+        connFactory = new LdapBoundConnFactory("LDAPRealm");
+        try {
+            connFactory.init(socketConfig, ldapConfig, ps);
+        } catch (Exception e) {
+            throw new LifecycleException("LDAP connection problem");
```

Comment:
Similarly, could you append the original exception message, and attach the original exception too?